### DISCUSSION TITLE
Can't access .env.local vars inside getStaticProps()

### DISCUSSION THREAD

**Discussion by nabeards** [vercel/next.js#31142] [Help]
### What version of Next.js are you using?

12.0.1

### What version of Node.js are you using?

14.15.1

### What browser are you using?

Safari

### What operating system are you using?

macOS

### How are you deploying your application?

next dev

### Describe the Bug

According to [Use .env.local to load environment variables](https://nextjs.org/docs/basic-features/environment-variables#loading-environment-variables), I should be able to access environment variables from `.env.local` inside `getStaticProps()`. This is not working, and neither is loading the `PORT` env var from `.env.local`. All env vars defined in `.env.local` are `undefined` when accessed from `getStaticProps()`.

I created this app through the Vercel web site, which created it inside my GitLab account. I then cloned it locally to customize.

### Expected Behavior

I expect `process.env.VARNAME` to be accessible within `getStaticProps()`.

### To Reproduce

.env.local:
```
PORT=3008
KEY=value1
PASS=value2
```

index.js:
```
export async function getStaticProps() {
    console.log('any are undefined:', process.env.KEY, process.env.PASS, process.env.PORT);
};
```

**Comment by timneutkens** [comment]
The code used to test it:
```jsx
export async function getStaticProps() {
  return {
    props: {
      KEY: process.env.KEY,
      PASS: process.env.PASS,
      PORT: process.env.PORT,
    },
  };
}

export default function Home(props) {
  return <pre>{JSON.stringify(props, null, "  ")}</pre>;
}
```


`KEY=1 PASS=234 PORT=8000 yarn dev`:

Result:
```
{
  "KEY": "1",
  "PASS": "234",
  "PORT": "8000"
}
```


---

Same but with `.env.local`:

```
KEY=1
PASS=234
PORT=8000
```

Result:
```
{
  "KEY": "1",
  "PASS": "234",
  "PORT": "8000"
}
```


Seems there's something in your app given that this is working as expected. I'll convert the issue to discussions.

**Reply by nabeards** [reply]
And you generated that code directly from the Vercel site as I described?

**Comment by nabeards** [comment]
Just did this from scratch. As you can see, it doesn't work:
https://nextjs-test-nine-tan.vercel.app

Here's the repo:
https://gitlab.com/nabeards/nextjs-test/

Something is wrong with what's available on the Vercel site, and it's wasted a few hours of my time at this point.

**Reply by nabeards** [reply]
Looks like my issue was prematurely closed.

**Comment by delbaoliveira** [comment]
Hey @nabeards 👋🏼 

Thanks for the link and the repo. Would you be able to confirm if you added the variables to your Vercel deployment? [Here's more information](https://vercel.com/docs/concepts/projects/environment-variables).

**Reply by nabeards** [reply]
That has nothing to do with my reported issue. Clone it local, create a .env.local with those vars, they are empty.

**Reply by nabeards** [reply]
I went ahead and added them to the production link, which works now. So you can't see how it fails locally until you clone it and add a `.env.local`...which is exactly what my original issue report states.

**Reply by delbaoliveira** [reply]
Glad it's working in production.

I've cloned Gitlab repo, created the `.env.local` file, added the environment variables and it's working locally. 

Could you help me understand what I'm missing?

**Reply by nabeards** [reply]
then I'm at a loss. you have that file in the root of the project? what OS do you test on? what permissions do you have on the `.env.local` file itself? there must be something different about my config than yours.

**Reply by delbaoliveira** [reply]
Root folder, macOS, nothing special about permissions:

```
ls -ld .env.local
-rw-r--r--  1 delba staff  24  8 Nov 20:40 .env.local
```

Also checked on another computer which worked. There is something we're doing that's different but I'm not sure what it is.

**Reply by nabeards** [reply]
Oy, ok, I had mine in the wrong folder. It was in `pages/`. I've now moved it to the project root, and it somewhat works. I get the expected output on screen:
```
{
  "KEY": "1",
  "PASS": "234",
  "PORT": "8000"
}
```

BUT, it's still running on the default PORT of 3000, not the assigned port of 8000. How do I get the next app to recognize that PORT I've set in the .env.local?

**Reply by delbaoliveira** [reply]
Glad it's working! Files inside the pages folder are considered [routes in Next.js](https://nextjs.org/docs/routing/introduction), so to avoid unexpected behaviour, it's best to keep .env or any config files outside the pages folder. 

For the port, setting it from a .env file is not possible because the server boots up before any code is initialized ([reference](https://nextjs.org/docs/api-reference/cli#development)).

This may not be a solution you're looking for, but perhaps worth mentioning, you can set the port using a -p flag, for example, `next dev -p 8000`.

**Reply by nabeards** [reply]
Why would the example from @timneutkens use a PORT env variable if it's completely useless?

**Reply by timneutkens** [reply]
I just took what you provided. I didn't create the example. `.env.local` is loaded after the server is booted for performance reasons, so you can't set `PORT` there because it's already booted.

**Reply by nabeards** [reply]
Okie doke! Appreciate the discussion.
